### PR TITLE
[RFC] Add metrics exporter interface

### PR DESCRIPTION
**Pull request by adleong**
Proof of concept.

We define a new plugin type `MetricsExporter` and add a config section for a list of metrics exporters.  At a configurable interval, a dump of the current metrics is given to each exporter.

### CODE REVIEW

File: linkerd/core/src/main/scala/io/buoyant/linkerd/Linker.scala [hunk 1]
```diff
@@ -74,7 +77,8 @@ object Linker {
     namers: Option[Seq[NamerConfig]],
     routers: Seq[RouterConfig],
     tracers: Option[Seq[TracerConfig]],
-    admin: Option[AdminConfig]
+    admin: Option[AdminConfig],
+    metrics: Option[Seq[MetricsExporterConfig]]
```
Comment (olix0r) [inline linkerd/core/src/main/scala/io/buoyant/linkerd/Linker.scala]: bikeshed:  It seems like a _metrics_ section might have some other things (like, configuration that controls how metrics are _recorded_ in addition to exported.  Does it make sense to structure this as:

```yaml
metrics:
  exporters:
  - ... 
```

File: linkerd/core/src/main/scala/io/buoyant/linkerd/MetricsExporterInitializer.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+trait MetricsExporterConfig {
+
+  @JsonProperty("periodSecs")
+  var _periodSecs: Option[Int] = None
+
+  @JsonIgnore
+  def period = _periodSecs.map(_.seconds).getOrElse(1.minute)
+
+  @JsonIgnore
+  def export(metrics: mutable.Map[String, Number]): Future[Unit]
```
Comment (olix0r) [inline linkerd/core/src/main/scala/io/buoyant/linkerd/MetricsExporterInitializer.scala]: I'd be inclined to separate configuration types from the type that actually does the work. I'd prefer to see something like `def mk(timer: Timer): MetricsExporter` here.

File: linkerd/core/src/main/scala/io/buoyant/linkerd/MetricsExport.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  exporters: Seq[MetricsExporterConfig],
+  metrics: MetricProvider = MetricsStatsReceiver.defaultRegistry,
+  timer: Timer = DefaultTimer.twitter
+) extends Closable {
+
+  private[this] val tasks = exporters.map { exporter =>
+    timer.schedule(exporter.period) {
+      val _ = exporter.export(metrics.sample().asScala)
+    }
+  }
```
Comment (olix0r) [inline linkerd/core/src/main/scala/io/buoyant/linkerd/MetricsExport.scala]: I'd be inclined to push this logic into each exporter.  There's nothing that says stats export has to be on a fixed interval that's known up front.  Let the exporter make that decision.  MetricsExporter could simply expose Closable.  Then, all exporters can be joined with `Closable.all(exporters:_*)`

Comment (olix0r) [inline linkerd/core/src/main/scala/io/buoyant/linkerd/MetricsExport.scala]: That said, it definitely seems like a good idea to provide PeriodicMetricsExporter and PeriodicMetricsExporterConfig types to make it easy to write these.

**Comment by adleong** [comment]
Closing in favor of https://github.com/BuoyantIO/linkerd/pull/580

**Pull request closed without merging**